### PR TITLE
Implement hash table using chaining technique

### LINKED ISSUE #1: Implement Hash Table with chaining using Doubly Linked List
Implement HashTable with chaining using Linked Lists. There are some main functions:
- insert(self, key, val)
- remove(self, key)
- search(self, key)

### PR DESCRIPTION
**Pull request by nvdai2401**
Close #1

### CODE REVIEW

**Comment by trungducc** [comment]
@vietduc01100001 please help to take a look

**Comment by trungducc** [comment]
@nvdai2401 test cases are missing. You need to add test cases for each method.

File: hash-table/chaining.js [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+          cur = cur.prev
+        }
+      } else {
+        this.arr[index].pushBack(value, key)
+      }
+    }
+
+    if (this.size === this.capacity) {
+      this.extendCapacity()
+    }
```
Comment (hellovietduc) [inline hash-table/chaining.js]: Can you do this part with less nesting if else?

File: hash-table/chaining.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+          break
+        }
+        count++
+        cur = cur.prev
+      }
+    } else {
+      this.arr[index] = undefined
+      this.size--
+    }
+    if (this.capacity > 8 && this.size <= Math.floor(this.capacity / 4)) {
```
Comment (hellovietduc) [inline hash-table/chaining.js]: We don't keep the magic numbers in the code, Mr @trungducc said :)) Grow/shrink factor, default capacity, etc, all should be declared as constants somewhere easy to change, so we don't have to go to each function and change them.

File: hash-table/chaining.js [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    }
+  }
+
+  search(key) {
+    if (!this.size) return null
+    const index = this.hash(key)
+    
+    if (!this.arr[index]) {
+      return null
+    } else {
```
Comment (hellovietduc) [inline hash-table/chaining.js]: `else` is not needed after a `return` statement.

File: hash-table/chaining.js [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        while (cur) {
+          linkList.pushBack(cur.value, cur.key)
+          cur = cur.next
+        }
+        newArr[index] = linkList
+      }
+    }
+    return newArr
+  }
+}
```
Comment (hellovietduc) [inline hash-table/chaining.js]: I think you misunderstand the point of growing/shrinking. For example, after growing capacity, `hash` function will return different results, since it `% capacity`. If you don't rehash, the items are now stored at the wrong indices in the bucket array. Or worse, they are still stored in the Linked Lists, which makes no sense since we now have more capacity but still have to loop through the Linked Lists to find the items.

Basically, changing capacity means rehashing all the items in the HashTable.

File: hash-table/chaining.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
+const { Node, DoublyLinkedList } = require('../linked-list/doubly-linked-list')
+
+class ChainingHashTable {
+  constructor(capacity = 8) {
+    this.arr = new Array(capacity)
```
Comment (hellovietduc) [inline hash-table/chaining.js]: What if I pass negative integers, or float numbers to this constructor? Or someone might pass a string as well (JS disadvantage, maybe)?

**Review by hellovietduc — CHANGES_REQUESTED** [review]

**Comment by nvdai2401** [comment]
> 
> 
> @nvdai2401 test cases are missing. You need to add test cases for each method.

@trungducc  I've implemented test cases for each method. Please check it again!

**Comment by nvdai2401** [comment]
@vietduc01100001 I've refactored my code base following your comments. Please check it again!

File: hash-table/chaining.js [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  }
+
+  delete(key) {
+    const index = this.hash(key)
+    
+    if (!this.arr[index]) return
+    if (this.arr[index].length === 1) {
+      this.arr[index] = undefined
+      this.size--
+      if (this.capacity > DEFAULT_CAPACITY && this.size <= Math.floor(this.capacity / 4)) {
```
Comment (hellovietduc) [inline hash-table/chaining.js]: There's still one more magic number here, which is 4. Ideally, you should check if the *load factor* of the Hash Table is less than a number, if yes then shrink. Load factor varies from 0 to 1. You can read more about it [here](https://en.wikipedia.org/wiki/Hash_table#Key_statistics).

File: hash-table/chaining.js [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+    while (cur) {
+      if (cur.key === key) {
+        return cur.value
+      }
+      cur = cur.prev
+    }
+  }
+
+  // private methods
+  hash(key) {
```
Comment (hellovietduc) [inline hash-table/chaining.js]: A common convention I see for marking private properties/methods in JS is to add `_` as prefix. If you know any cooler tricks, tell me. :)

File: hash-table/chaining.js [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+    return key % this.capacity
+  }
+
+  grow() {
+    this.capacity *= 2
```
Comment (hellovietduc) [inline hash-table/chaining.js]: Grow factor `2` should be a constant, not hard-coded.

File: hash-table/chaining.js [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+        }
+      } while (length < this.capacity)
+    }
+    this.arr = newArr
+    this.size = length
+  }
+
+  shrink() {
+    const oldCapacity = this.capacity
+    this.capacity /= 2
```
Comment (hellovietduc) [inline hash-table/chaining.js]: Shrink factor should also not be hard-coded.

File: hash-table/chaining.js [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+          newArr[index] = linkList
+          length++
+          cur = cur.next
+        }
+      } 
+    }
+
+    return { newArr, length }
+  }
+}
```
Comment (hellovietduc) [inline hash-table/chaining.js]: Here's a simpler way you can do this:
1. Create a new array with capacity calculated by a grow/shrink factor.
2. Create a new variable which points to the current bucket array.
3. Assign the new array as the bucket array of the Hash Table.
4. Loop through all the elements in your old bucket array, including ones in the Linked Lists.
5. For each element, call the `insert()` function. It should re-compute the hash and insert the elements into the new bucket array.

You shouldn't need to create Linked Lists and call the `hash` function manually. Your `insert()` function already does that.

**Review by hellovietduc — CHANGES_REQUESTED** [review]

**Comment by hellovietduc** [comment]
> @vietduc01100001 I've refactored my code base following your comments. Please check it again!

You can resolve my outdated comments once you solved that.
